src/ClientData/CaptureData.cpp

```diff
@@ -251,4 +252,31 @@ void CaptureData::UpdateTimerDurations() {
   return result;
 }
 
+[[nodiscard]] std::optional<ThreadStateSliceInfo>
+CaptureData::FindThreadStateSliceInfoFromTimestamp(int64_t thread_id, uint64_t timestamp) const {
+  absl::MutexLock lock{&thread_state_slices_mutex_};
+  if (!thread_state_slices_.contains(thread_id)) {
+    return std::nullopt;
+  }
+
+  const std::vector<ThreadStateSliceInfo>& thread_state_bar = thread_state_slices_.at(thread_id);
+  auto slice =
+      std::upper_bound(  // This should return the slice directly in front of the slice that we want
+          thread_state_bar.begin(), thread_state_bar.end(), timestamp,
+          [](uint64_t b,
+             const ThreadStateSliceInfo& a) -> bool {  // compare based on beginning timestamps
+            return b < a.begin_timestamp_ns();
```

Comment:
If you compare with end_timestamp_ns(), you can make this method much simpler. The UpperBound will reply the exact interval that you want.

After you just need to compare if it's not end() (after the last one) and also timestamp < slice->begin_timestamp_ns() (that it's not before the first one)